### DISCUSSION TITLE
Is it possible to extend a system interface?

### DISCUSSION THREAD

**Discussion by mattiaz9** [directus/directus#12377] [Archive]
It'd be nice to extend an existing directus interfaces to add more features without the need to rewrite the entire interface from scratch.
For instance, for the Markdown interface, inserting some buttons in the toolbar to add custom functionalities. 

I know Vue has some extend functionalities, but I'm not sure if it's possibile to use it for this scope or how to do it.

**Comment by rijkvanzanten** [comment]
We don't have a built-in way to do so, but I'm more than happy to keep discussing how that might work. With the knowledge that an interface is "just" a Vue component, how would you imagine this extending would work?

From a Vue perspective, extending is effectively reusing mixins, which as far as I'm aware is purely functional logic (eg methods) rather than markup. In the case of extending the markdown header, you would have to have some way to extend a very particular part of the markup used in the component, without copying any/most of the existing interface 🤔 

I think the only way to achieve such a thing would be for us to explicitly mark what parts of an interface can be extended with custom code through Slots for example, but that would be a tremendous amount of work to maintain 🤔

**Reply by benhaynes** [reply]
My question is in response to:

> "...rewrite the entire interface from scratch..."

Extending interfaces is a very difficult request, but my assumption was that any core interface/module can simply be duplicated, tweaked, and given a new custom name. If that is true, then building a _slightly_ different interface/layout isn't about rebuilding, it's just about forking.

Is there some core logic embedded in our core interfaces that makes this not possible?

**Reply by rijkvanzanten** [reply]
Most interfaces can be reproduced easily in that way. The main issue with doing that is that there's no real way to easily keep your custom interface fork up to date with any fixes/improvements that we make in the "real" one. Having a way to inject a tiny customization without touching the rest would prevent that problem 🤔

**Reply by mattiaz9** [reply]
@benhaynes @rijkvanzanten  Thanks for the replies. I understand that this would be quite difficult to implement.

And yes, I 'forked' the markdown interface, but I wanted to make sure if there was an easiest and more elegant way to extend it. Especially because I have also to include some big dependencies, like code mirror that are already used in the original one.

Also some other interfaces a wanted to extend are the the `list-m2o`, `list-m2m` ext.. to allow them to resolve the translations, since right now they only show dashes.

**Reply by MichaelFBA** [reply]
I noticed there is the `useExtensions` method available on the extensions-sdk, is there a possibility to somehow use this to inject the extension into a new component?
I was also looking at updating the translations interface

**Reply by rijkvanzanten** [reply]
No, that's not meant to extend an existing core interface 👍🏻

**Reply by MichaelFBA** [reply]
I was able to extend it in a simple fashion like so. If it helps anyone
```
<script lang="ts">
import { useExtensions } from "@directus/extensions-sdk";
import { defineComponent } from "@vue/runtime-core";

export default defineComponent({
	setup(props, {attrs, emit}) {
		const { interfaces } = useExtensions();
		const { component: TranslationComponent } = interfaces.value.find(
		(i) => i.id === "translations"
		);
		return {
			attrs,
			TranslationComponent,
		};
	}
})
</script>

<template>
  <component :is="TranslationComponent" v-bind="attrs" />
  <h1>My Custom additions<h1>
</template>
```

**Reply by jay-p-b-7span** [reply]
Thank you for sharing. it works great. Your solution has saved lot of time 😊

**Comment by mallle** [comment]
Hi @MichaelFBA - I wonder if you would have a moment and maybe be willing to help me with a  I have a question ?

I'm rendering the `input-rich-text-md` component in the same way as you above with the TranslationComponent. But I have some trouble retreving the changes from the field. 

I have tried to do something like this: 
```
 <component :is="MarkdownComponent" :bind="attrs" :value="text" @keyup="handleChange($event.target.value, language)"/>
```
And then I want to emit the changes inside the `handleChange`.
```
function handleChange(newValue, transferLang) {
  informationPair.value[transferLang] = newValue;
  const string = JSON.stringify(informationPair.value);
  emit("handle-input", string);
}
```
`newValue` is null, so this is definitely not the way to do it. Would you be able to help me?

**Reply by MichaelFBA** [reply]
What about if you simply do this 
`@keyup="handleChange"`

You could also check the events emitted from the child component. Perhaps you could use v-model?

**Reply by mallle** [reply]
Thanks for the quick response.
`@keyup="handleChange"`, does give me the changes, but I still need the informations from the second variable that I'm passing, (the `language`).  Do you have any idea on how I could do that? 
Im rendering mulitple markdown components in a for loop, and therefore need the `language` to determine where in the loop the value is getting changed:
```
 <component :is="MarkdownComponent" :bind="attrs" :value="text" @keyup="handleChange($event.target.value, language)"/>

```

**Reply by mallle** [reply]
Ohh i just had to call `@input="handleChange($event, language)`. Thanks a lot 👍

**Comment by miguelrs** [comment]
@MichaelFBA @mallle I've tried that approach you mentioned and it does work to add some extra markup for example 👍 

However, what I wanted to do is _modifying_ the markup of a more complicated component, like [select-dropdown-m2o.vue](https://github.com/directus/directus/blob/main/app/src/interfaces/select-dropdown-m2o/select-dropdown-m2o.vue). For example, I'd like to do something apparently simple like removing the "edit" icon that appears when an item is selected and opens the drawer for editing, which is just this markup within the template:
```
<v-icon v-tooltip="t('edit')" name="open_in_new" class="edit" @click.stop="editModalActive = true" />
```

Do you think is it possible to do something like this? 

Sorry I'm absolutely new to VueJS, so I'm just trying things and don't know what I'm doing. Maybe you can point me in the right direction, or at least tell me it's not possible so I don't waste more time 🙏😅

**Comment by MichaelFBA** [comment]
@miguelrs I think it's not easy to do at the moment, you would probably need in effect fork/copy the component and modify it to your needs. 

That has been my experience so far

**Reply by miguelrs** [reply]
@MichaelFBA Thanks a lot for your response! ❤️ 

Just one thing if you have a sec: when you all say "fork/copy the component" you literally mean going to the `directus/directus` repo, copy the `.vue` file of the component along with all the `.ts` files required, paste them into my extended interface repo (e.g. `miguelrs/my-custom-interface`), build the extension and copy the `index.js` into my personal directus project (e.g. `miguelrs/my-directus-app`), right?

So for example, for the [select-dropdown-m2o.vue](https://github.com/directus/directus/blob/main/app/src/interfaces/select-dropdown-m2o/select-dropdown-m2o.vue), I have to go through the `import` statements in its `.vue` file, and copy&paste all the required files, i.e:
```
// I know how to get these from @directus/extensions-sdk:
import { usePermissionsStore } from '@/stores/permissions';
import { useCollectionsStore } from '@/stores/collections';
import { Filter } from '@directus/shared/types';
import { deepMap, getFieldsFromTemplate } from '@directus/shared/utils';

// I know how to get these from the relevant libraries:
import { get } from 'lodash';
import { render } from 'micromustache';
import { computed, inject, ref, toRefs } from 'vue';
import { useI18n } from 'vue-i18n';

// But all these need to copy & pasted as well:
import { RelationQuerySingle, useRelationSingle } from '@/composables/use-relation-single';
import { useRelationM2O } from '@/composables/use-relation-m2o';
import { adjustFieldsForDisplays } from '@/utils/adjust-fields-for-displays';
import { parseFilter } from '@/utils/parse-filter';
import DrawerCollection from '@/views/private/components/drawer-collection.vue';
import DrawerItem from '@/views/private/components/drawer-item.vue';
```

So I have to copy&paste at least 4 or 5 files, which in turn have more imports I don't have access to, so I have to copy&paste more files, etc... 

That's how it is, isn't it? Or am I missing something?

--- 

An alternative approach I'm thinking of, is not using `npm init directus-project miguelrs/my-directus-app` to create my project, but instead forking the whole `directus/directus` repo, remove the `.git` folder, and use that repo as my project, so I have direct access to all the components, and I can just make my small changes there. I've started trying this and seems I can do my tweaks easily! Not sure I'll encounter other issues later on 🤷 Does this approach make any sense?

**Reply by MichaelFBA** [reply]
You would need to replace most of the imports with the extensions/SDK provided by directus. They provide `useStores` etc 

The alternative you mentioned is basically forking the whole project which would diverge over time, that's why directus provides extensions to be able to extend the platform.

**Reply by AxeemHaider** [reply]
@MichaelFBA But you can't import these `@/composables/use-relation-single` kind of stuff from '@directus/extensions-sdk` how you resolve these kind of **imports** ?

**Reply by MichaelFBA** [reply]
![Screenshot 2023-03-20 at 08 57 04](https://user-images.githubusercontent.com/1113721/226279294-405717da-10e5-4071-8979-58d8369dc3a6.png)

`useStores()` provides some functionality to reproduce
